radix-engine/src/engine/actor.rs

```diff
@@ -36,69 +68,25 @@ pub enum REActor {
-                ) => true,
-                (RENodeId::Global(_), SubstateOffset::Global(GlobalOffset::Global)) => true,
-                (RENodeId::Component(_), SubstateOffset::Component(ComponentOffset::Info)) => true,
-                _ => false,
-            },
+    pub fn is_scrypto_or_transaction(&self) -> bool {
+        matches!(
+            self,
             REActor::Method(ResolvedReceiverMethod {
-                receiver: Receiver::Ref(RENodeId::Component(component_address)),
                 method: ResolvedMethod::Scrypto { .. },
-            }) => match (node_id, offset) {
-                (
-                    RENodeId::KeyValueStore(_),
-                    SubstateOffset::KeyValueStore(KeyValueStoreOffset::Entry(..)),
-                ) => true,
-                (RENodeId::Component(_), SubstateOffset::Component(ComponentOffset::Info)) => true,
-                (RENodeId::Component(addr), SubstateOffset::Component(ComponentOffset::State)) => {
-                    addr.eq(component_address)
-                }
-                (RENodeId::Global(_), SubstateOffset::Global(GlobalOffset::Global)) => true,
-                _ => false,
-            },
-            _ => false,
-        }
-    }
-
-    pub fn is_substate_writeable(&self, node_id: RENodeId, offset: SubstateOffset) -> bool {
-        match self {
-            REActor::Function(ResolvedFunction::Native(..))
-            | REActor::Method(ResolvedReceiverMethod {
-                method: ResolvedMethod::Native(..),
                 ..
-            }) => true,
-            REActor::Function(ResolvedFunction::Scrypto { .. }) => match (node_id, offset) {
-                (
-                    RENodeId::KeyValueStore(_),
-                    SubstateOffset::KeyValueStore(KeyValueStoreOffset::Entry(..)),
-                ) => true,
-                _ => false,
-            },
-            REActor::Method(ResolvedReceiverMethod {
-                receiver: Receiver::Ref(RENodeId::Component(component_address)),
-                method: ResolvedMethod::Scrypto { .. },
-            }) => match (node_id, offset) {
-                (
-                    RENodeId::KeyValueStore(_),
-                    SubstateOffset::KeyValueStore(KeyValueStoreOffset::Entry(..)),
-                ) => true,
-                (RENodeId::Component(addr), SubstateOffset::Component(ComponentOffset::State)) => {
-                    addr.eq(component_address)
-                }
-                _ => false,
-            },
-            _ => false,
-        }
+            }) | REActor::Function(ResolvedFunction::Scrypto { .. })
+                | REActor::Function(ResolvedFunction::Native(
+                    NativeFunction::TransactionProcessor(TransactionProcessorFunction::Run)
+                ))
+        )
     }
 }
+
+/// Executing kernel level actor
+#[derive(Debug, Copy, Clone, Eq, PartialEq, TypeId, Encode, Decode)]
+pub enum KernelActor {
```

Comment:
Actor is being overloaded :)

Suggested renaming to `RestrictedMode` which further restricted what system APIs are allowed.
```rust
pub enum RestrictedMode {
   Deref,
   ScryptoFnResolver,
   AuthModule,
   /* possibly other kernel modules */
}
```